haystack/telemetry.py

```diff
@@ -0,0 +1,343 @@
+            config = yaml.safe_load(stream)
+            if "user_id" in config and user_id is None:
+                user_id = config["user_id"]
+    except Exception:
+        pass
+
+
+def _write_telemetry_config():
+    """
+    Writes a config file storing the randomly generated user id and whether to write events to a log file.
+    This method logs an info to inform the user about telemetry when it is used for the first time.
+    """
+    global user_id
+    try:
+        # show a log message if telemetry config is written for the first time
+        if not Path(CONFIG_PATH).is_file():
+            logger.info(
+                f'Haystack sends anonymous usage data to understand the actual usage and steer dev efforts towards features that are most meaningful to users. You can opt out at anytime by setting {HAYSTACK_TELEMETRY_ENABLED}="False" as an environment variable or by calling disable_telemetry(). More information at https://haystack.deepset.ai/overview/faq'
+            )
+            Path(CONFIG_PATH).parents[0].mkdir(parents=True, exist_ok=True)
+        user_id = _get_or_create_user_id()
+        config = {"user_id": user_id}
+
+        with open(CONFIG_PATH, "w") as outfile:
+            yaml.dump(config, outfile, default_flow_style=False)
+    except Exception:
+        send_custom_event(event="config saving failed")
+
+
+def _delete_telemetry_config():
+    """
+    Deletes the telemetry config file if any exists.
+    """
+    try:
+        Path(CONFIG_PATH).unlink(missing_ok=True)
+    except Exception:
+        pass
+
+
+def _write_event_to_telemetry_log_file(distinct_id: str, event: str, properties: Dict[str, Any]):
+    try:
+        with open(LOG_PATH, "a") as file_object:
+            file_object.write(f"{event}, {properties}, {distinct_id}\n")
+    except Exception:
+        pass
+
+
+def _delete_telemetry_log_file():
+    """
+    Deletes the telemetry log file if any exists.
+    """
+    try:
+        Path(LOG_PATH).unlink(missing_ok=True)
+    except Exception:
+        pass
+
+
+class NonPrivateParameters:
+    param_names: List[str] = ["top_k", "model_name_or_path", "add_isolated_node_eval"]
+
+    @classmethod
+    def apply_filter(cls, param_dicts: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Ensures that only the values of non-private parameters are sent in events. All other parameter values are filtered out before sending an event.
+        If model_name_or_path is a local file path, it will be reduced to the name of the file. The directory names are not sent.
+
+        :param param_dicts: the keyword arguments that need to be filtered before sending an event
+        """
+        tracked_params = {k: param_dicts[k] for k in cls.param_names if k in param_dicts}
+
+        # if model_name_or_path is a local file path, we reduce it to the model name
+        if "model_name_or_path" in tracked_params:
+            if Path(tracked_params["model_name_or_path"]).is_file():
```

Comment:
The `is_file()` check is to handle the case of private models that are stored locally. It's not for cached models from the model hub. Private local models should exist before a node calls `super().__init__()`. They are not downloaded.
For non-existing files I can add a check that if `model_name_or_path` contains more than one `/` we assume it is a path and shorten it to the filename. That will handle the case that a local relative path with more than one `/` points to a file that does not exist (yet). 
What remains as a tiny problem is a Path containing only one `/` pointing to a non-existing file, e.g., because of a typo. In that case, that name would be sent in an event.